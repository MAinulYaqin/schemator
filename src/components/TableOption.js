import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

import "typeface-roboto";

const Container = styled.div`
  width: 100%;
  height: 72px;
  padding: 0px 8px;
  box-sizing: border-box;
  background: #222;
`;

const Row = styled.div`
  display: flex;
  align-items: center;
  height: 36px;
`;

const Label = styled.label`
  font-family: Roboto;
  font-size: 12px;
  color: white;
  margin-right: 12px;
`;

export const Input = styled.input`
  outline: none;
`;

const TableOption = ({ value, onChange }) => (
  <Container>
    <Row>
      <Input
        type="checkbox"
        checked={value.id}
        onChange={event => onChange(event, "id")}
      />
      <Label>ID</Label>
      <Input
        type="checkbox"
        checked={value.rememberToken}
        onChange={event => onChange(event, "rememberToken")}
      />
      <Label>Remember Token</Label>
    </Row>
    <Row>
      <Input
        type="checkbox"
        checked={value.softDeletes}
        onChange={event => onChange(event, "softDeletes")}
      />
      <Label>Soft Deletes</Label>
      <Input
        type="checkbox"
        checked={value.timestamps}
        onChange={event => onChange(event, "timestamps")}
      />
      <Label>Timestamps</Label>
    </Row>
  </Container>
);

TableOption.propTypes = {
  value: PropTypes.object,
  onChange: PropTypes.func
};

TableOption.defaultProps = {
  value: {
    id: false,
    rememberToken: false,
    softDeletes: false,
    timestamps: false
  }
};

export default TableOption;
