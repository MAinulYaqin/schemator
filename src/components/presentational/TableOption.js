import React, { Component } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

import "typeface-roboto";

const Container = styled.div`
  width: 100%;
  height: 36px;
  padding: 0px 8px;
  box-sizing: border-box;
  background: #222;
  display: flex;
  align-items: center;
`;

export const Item = styled.div`
  display: flex;
  align-items: center;
  margin-right: ${({ isFirst }) => (isFirst ? 16 : 0)}px;
`;

const Label = styled.label`
  font-family: Roboto;
  font-size: 12px;
  color: white;
`;

export const Input = styled.input`
  outline: none;
  margin-right: 4px;
`;

class TableOption extends Component {
  shouldComponentUpdate(nextProps) {
    const { items } = this.props;

    const diffs = items.map((item, index) => {
      const nextItem = nextProps.items[index];
      const isCheckedDiff = nextItem.isChecked !== item.isChecked;

      return isCheckedDiff;
    });

    return diffs.includes(true);
  }

  render() {
    const { items, onChange } = this.props;

    return (
      <Container>
        {items.slice(0, 2).map((item, index) => (
          <Item key={item.id} isFirst={index === 0}>
            <Input
              type="checkbox"
              checked={item.isChecked}
              onChange={event => onChange(event, item.id)}
            />
            <Label>{item.label}</Label>
          </Item>
        ))}
      </Container>
    );
  }
}

TableOption.propTypes = {
  items: PropTypes.array,
  onChange: PropTypes.func
};

TableOption.defaultProps = {
  items: []
};

export default TableOption;
