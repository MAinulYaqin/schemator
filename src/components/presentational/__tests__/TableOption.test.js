import React from "react";
import { shallow } from "enzyme";
import "jest-styled-components";

import TableOption, { Item, Input } from "../TableOption";

describe("TableOption", () => {
  const setup = propOverrides => {
    const props = {
      items: [
        { id: "id", label: "ID", isChecked: true },
        { id: "rememberToken", label: "Remember Token", isChecked: false }
      ],
      onChange: jest.fn(),
      ...propOverrides
    };

    const wrapper = shallow(<TableOption {...props} />);

    return {
      wrapper,
      props
    };
  };

  it("should renders default", () => {
    const { wrapper } = setup();
    const item = wrapper.find(Item);

    expect(item.at(0)).toHaveStyleRule("margin-right", "16px");
    expect(item.at(1)).toHaveStyleRule("margin-right", "0px");
    expect(wrapper).toMatchSnapshot();
  });

  it("should calls 'onChange' while changed", () => {
    const { wrapper, props } = setup();
    const optionIDs = ["id", "rememberToken"];

    const event = {
      target: {
        checked: true
      }
    };

    optionIDs.forEach((item, index) => {
      wrapper
        .find(Input)
        .at(index)
        .simulate("change", event, item);

      expect(props.onChange).toBeCalledWith(event, item);
    });
  });
});
