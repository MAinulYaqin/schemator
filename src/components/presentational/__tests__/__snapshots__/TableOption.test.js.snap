// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableOption should renders default 1`] = `
<styled.div>
  <styled.div
    isFirst={true}
    key="id"
  >
    <styled.input
      checked={true}
      onChange={[Function]}
      type="checkbox"
    />
    <styled.label>
      ID
    </styled.label>
  </styled.div>
  <styled.div
    isFirst={false}
    key="rememberToken"
  >
    <styled.input
      checked={false}
      onChange={[Function]}
      type="checkbox"
    />
    <styled.label>
      Remember Token
    </styled.label>
  </styled.div>
</styled.div>
`;
